tests/auxiliary/user-account/test_contract.py

```diff
@@ -0,0 +1,372 @@
+
+from eth_keys import keys
+
+from evm.constants import (
+    UINT_256_MAX,
+    SECPK1_N,
+)
+from evm.vm.forks.sharding import (
+    ShardingVM,
+)
+from evm.vm.forks.sharding.transactions import (
+    ShardingTransaction,
+)
+
+from evm.rlp.headers import (
+    BlockHeader,
+)
+
+from evm.db import (
+    get_db_backend,
+)
+from evm.db.chain import (
+    BaseChainDB,
+)
+from evm.db.state import (
+    FlatTrieBackend,
+    NestedTrieBackend,
+)
+
+from eth_utils import (
+    to_canonical_address,
+    decode_hex,
+    int_to_big_endian,
+    big_endian_to_int
+)
+from evm.utils.padding import pad32
+from evm.utils.address import (
+    generate_create2_contract_address,
+)
+from evm.utils.keccak import keccak
+
+from evm.auxiliary.user_account_contract.transaction import (
+    ForwardingTransaction,
+)
+from evm.auxiliary.user_account_contract.contract import (
+    generate_validation_bytecode,
+    generate_account_bytecode,
+    NONCE_GETTER_ID,
+    VALIDATION_CODE_GETTER_ID,
+)
+
+
+PRIVATE_KEY = keys.PrivateKey(b"\x33" * 32)
+
+VALIDATION_CODE = generate_validation_bytecode(PRIVATE_KEY.public_key.to_canonical_address())
+VALIDATION_CODE_ADDRESS = generate_create2_contract_address(b"", VALIDATION_CODE)
+
+ACCOUNT_CODE = generate_account_bytecode(VALIDATION_CODE_ADDRESS)
+ACCOUNT_ADDRESS = generate_create2_contract_address(b"", ACCOUNT_CODE)
+INITIAL_BALANCE = 10000000000
+
+# contract that does nothing
+NOOP_CONTRACT_CODE = b""
+NOOP_CONTRACT_ADDRESS = generate_create2_contract_address(b"", NOOP_CONTRACT_CODE)
+
+# contract that reverts without returning data
+FAILING_CONTRACT_CODE = b"\x61\x00\x00\xfd"  # PUSH2 0 0 REVERT
+FAILING_CONTRACT_ADDRESS = generate_create2_contract_address(b"", FAILING_CONTRACT_CODE)
+
+# contract that logs available gas
+# GAS PUSH1 0 MSTORE PUSH1 32 PUSH1 0 LOG0
+GAS_LOGGING_CONTRACT_CODE = b"\x5a\x60\x00\x52\x60\x20\x60\x00\xa0"
+GAS_LOGGING_CONTRACT_ADDRESS = generate_create2_contract_address(b"", GAS_LOGGING_CONTRACT_CODE)
+
+# contract that logs hash of passed data
+# CALLDATASIZE PUSH1 0 PUSH1 0 CALLDATACOPY CALLDATASIZE PUSH1 0 SHA3 PUSH1 0 MSTORE PUSH1 32
+# PUSH1 0 LOG0
+DATA_LOGGING_CONTRACT_CODE = (
+    b"\x36\x60\x00\x60\x00\x37\x36\x60\x00\x20\x60\x00\x52\x60\x20\x60\x00\xa0"
+)
+DATA_LOGGING_CONTRACT_ADDRESS = generate_create2_contract_address(b"", DATA_LOGGING_CONTRACT_CODE)
+
+HELPER_CONTRACTS = {
+    VALIDATION_CODE_ADDRESS: VALIDATION_CODE,
+    ACCOUNT_ADDRESS: ACCOUNT_CODE,
+    NOOP_CONTRACT_ADDRESS: NOOP_CONTRACT_CODE,
+    FAILING_CONTRACT_ADDRESS: FAILING_CONTRACT_CODE,
+    GAS_LOGGING_CONTRACT_ADDRESS: GAS_LOGGING_CONTRACT_CODE,
+    DATA_LOGGING_CONTRACT_ADDRESS: DATA_LOGGING_CONTRACT_CODE,
+}
+
+
+DESTINATION_ADDRESS = b"\xbb" * 20
+
+DEFAULT_BASE_TX_PARAMS = {
+    "chain_id": 1,
+    "shard_id": 1,
+    "to": ACCOUNT_ADDRESS,
+    "gas": 500000,
+    "gas_price": 0,
+    "access_list": [],
+    "code": b"",
+}
+
+DEFAULT_TX_PARAMS = {**DEFAULT_BASE_TX_PARAMS, **{
```

Comment:
This method of dictionary merging has always bugged me, mostly because it takes advantage of the implementation rather than using formal semantics or which keys take precedence.  I admit it could be argued that the way key precedence is handled in this case is part of the formal semantics for how this works, but either way, it's always felt a bit like magic.

What do you think about making use of something like the following.  (If you haven't figured it out by now, I have great love for the `toolz/cytoolz` library)

```python
from cytoolz import (
    merge,
    dissoc,
)

DEFAUT_TX_PARAMS = merge(
    dissoc(DEFAULT_BASE_TX_PARAMS, 'code'),
    {...<the overrides>},
}
```

Zero mutation and explicit about dictionary merging.

Reply:
Personally I don't prefer one way over the other, so I don't mind adopting this style.